src/sharing/SharingDialog.component.js

```diff
@@ -1,44 +1,251 @@
+        access => transformAccessObject(access, 'user'));
+
+    const combinedAccesses = userGroupAccesses.concat(userAccesses);
+
+    return {
+        ...apiObject.user && {
+            authorOfSharableItem: {
+                id: apiObject.user.id,
+                name: apiObject.user.name,
+            },
+        },
+        nameOfSharableItem: apiObject.name,
+        canSetPublicAccess: apiMeta.allowPublicAccess,
+        canSetExternalAccess: apiMeta.allowExternalAccess,
+        publicCanView: apiObject.publicAccess.includes('r'),
+        publicCanEdit: apiObject.publicAccess.includes('rw'),
+        isSharedExternally: apiObject.externalAccess,
+        accesses: combinedAccesses,
+    };
+}
+
+/**
+ * A pop-up dialog for changing sharing preferences for a sharable object.
+ */
+class SharingDialog extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            apiObject: null,
+            objectToShare: null,
+        };
+
+        this.loadObjectFromApi();
+    }
+
+    componentWillReceiveProps(nextProps) {
+        if (nextProps.id !== this.props.id) {
+            this.setState({
+                objectToShare: null,
+            });
+
+            this.loadObjectFromApi();
+        }
+    }
+
+    onSearchRequest(searchText) {
+        const apiInstance = this.state.api;
+
+        return apiInstance.get('sharing/search', { key: searchText })
+            .then((searchResult) => {
+                let transformedResult = searchResult.users.map(
+                    user => transformAccessObject(user, 'user'));
+
+                transformedResult = transformedResult.concat(
+                    searchResult.userGroups.map(
+                        userGroup => transformAccessObject(userGroup, 'userGroup')));
+
+                return transformedResult;
+            });
+    }
+
+    onSharingChanged(updatedAttributes, onSuccess) {
+        const objectToShare = {
+            ...this.state.objectToShare,
+            ...updatedAttributes,
+        };
+
+        const apiObject = this.restoreObjectStructure(objectToShare);
+
+        return this.state.api.post(`sharing?type=${this.props.type}&id=${this.props.id}`, apiObject)
+          .then(({ httpStatus, message }) => {
+              if (httpStatus === 'OK') {
+                  this.setState({
+                      objectToShare,
+                      apiObject,
+                  }, () => {
+                      onSuccess && onSuccess();
+                  });
+              } else {
+                  console.warn('Failed to post changes.');
+                  console.warn('SERVER SAID:', message);
+              }
 
-    mixins: [Translate],
+              return message;
+          })
+          .catch(({ message }) => {
+              console.warn('Failed to post changes.');
+              console.warn('SERVER SAID:', message);
+          });
+    }
+
+    loadObjectFromApi() {
+        getInstance().then((d2) => {
```

Comment:
We could grab this one from the context. I think i might have suggested you to do it like this but sicne we already put `d2` on the context here we might aswel grab it from there.